manager/allocator/network/port/doc.go

```diff
@@ -0,0 +1,33 @@
+package port
+
+// The `port` package contains the implementation of the swarmkit Allocator for
+// network ports. Its purpose is to assign ports to Endpoint objects based on
+// the provided Spec. For ingress ports, it keeps track of the availability of
+// ports, and handles dynamic allocation of ports when the user has specified
+// no PublishedPort. The allocator does _not_ keep track of host-mode ports in
+// any capacity. It silently adds them to the returned Ports assignment,
+// without checking their validity or availability. Port assignments for
+// host-mode ports are handled on the worker.
+//
+// Notably, the allocator only keeps track of which ports are in use, not which
+// services they are in use by. For the port allocator to function correctly,
+// the objects it works with must be in a consistent state. To facilitate this,
+// the port allocator "owns" the Endpoint.Ports field, and should be the only
+// component that creates the PortConfig objects contained in it.
+//
+// Before the Allocator can be used, it needs to be initialized by calling its
+// Restore method. This populates the state of the Allocator with all of the
+// endpoints in use.
+//
+// The Allocator follows a two-phase transaction-style lifecycle. Calls to
+// Allocate or Deallocate return a `Proposal` object, which contains the
+// necessary information to commit the change. Calling the `Commit` method on
+// the Proposal will commit the changes to the Allocator. If the caller decides
+// to abandon the changes, the user can simply abandon the Proposal without
+// calling Commit.
```

Comment:
we can't support concurrency and rolling back. if we support concurrency, you may not be able to roll back a port allocation, because one of the ports you released may have been claimed. you'd have to split out allocation and de-allocation into separate steps.

if you look in the manager/allocator/network.go AllocateService method, we allocate ports, get a proposal, and then try allocation IP addresses. If the IP address allocation fails, we discard the proposal. in the ipam allocator, when allocation IP addresses, you allocate first, and then if allocation is successful, you free addresses. at that point, no further errors can occur, so you can safely commit the proposal. if you allocated ports after IP addresses, you might find that IP allocation succeeds but port allocation fails, and then you'd have to roll back IP address allocation in the same way and risk a freed resource being taken.

i think that this actually makes me realize there is one use for the `IsNoop` method on proposals (or some similar construct). concurrently allocating IP addresses would be a big win, but is blocked because a lock on the portallocator would have to be acquired before allocating and released after committing, and the IPAM call is inside that synchronized path. however, if you found the proposal was a noop and did not alter port state (which i suspect to be a more common case), you could release that lock immediately and allow other service allocations to proceed calling into the ip allocator concurrently.

on the other hand, i strongly suspect network updates to services are going to be infrequent enough to not substantially impact the performance of the allocator even if that whole code path is single-threaded.

Reply:
As stated by dperny the _commit inmediately and roll back later_ approach is not thread-safe for deallocations. The _prepare a proposal and commit later_ approach could be made thread-safe by keeping a logical clock in the allocator that gets updated every time a proposal is committed, but this would mean that any concurrent proposal would have to test its validity again. If a proposal is not valid due to a concurrent change then all the optimization of concurrency would be lost. And any deallocation in a proposal would make it non-valid, as you can not be sure if the committed proposal that triggered this validity check has deallocated and reallocated this resource.

Would an hybrid solution work? Allocating inmediatelly with roll back support, but not deallocating until committed. This will enforce the proposals to be either commited or reversed, dropping a proposal would result into some resources to be allocated forever until the swarm was rebuilt. I think this approach may be thread-safe, but it has to be taken into account that proposals hold both the resources they want to allocate and the ones that they want to deallocate, so a call to Allocate may seem not to have enough resources due to a concurrent proposal when it actually has.

Reply:
As stated by dperny the _commit inmediately and roll back later_ approach is not thread-safe for deallocations. The _prepare a proposal and commit later_ approach could be made thread-safe by keeping a logical clock in the allocator that gets updated every time a proposal is committed, but this would mean that any concurrent proposal would have to test its validity again. If a proposal is not valid due to a concurrent change then all the optimization of concurrency would be lost. And any deallocation in a proposal would make it non-valid, as you can not be sure if the committed proposal that triggered this validity check has deallocated and reallocated this resource.

Would an hybrid solution work? Allocating inmediatelly with roll back support, but not deallocating until committed. This will enforce the proposals to be either commited or reversed, dropping a proposal would result into some resources to be allocated forever until the swarm was rebuilt. I think this approach may be thread-safe, but it has to be taken into account that proposals hold both the resources they want to allocate and the ones that they want to deallocate, so a call to Allocate may seem not to have enough resources due to a concurrent proposal when it actually has.

Reply:
I'm not sure if I understand what you mean by roll forward and commit but I don't think deallocs are safe in the _commit immediately and roll back later_ approach. Let me describe the proposed approaches to make sure we understand the same:

Resources need to be allocated and deallocated. To be thread safe this operations need to be done in an atomic way, this is, all of them need to be commited simultaneously or you can not guarantee that the laters will not fail. Three aproaches were mentioned:
1. _Prepare a proposal and commit later_: when preparing a proposal you can ensure its validity, but as the allocator state can change before commiting, the validity of the proposal can not be ensured at commit time. To ensure both we could use a logical clock on the state that would be updated with every commit, if the clock has different value at commit time that when the proposal was created, the proposal would need to be checked again, and if there was any conflict, not only a new proposal would be needed but also all the steps between the proposal generation and the commit time would need to be executed again. Deallocs in this case will never result in a conflict as deallocating is idempotent, and actually should never happen as despite the allocator not knowing who owns which resource, they have a single owner so only it will try to deallocate it. On the other hand, allocs can result in conflicts if that same resource was already allocated (and commited) concurrently.
2. _Commit immediately and roll back later_: the oposite aproach, commit at proposal preparation and rollback if unsuccessful. A stright difference from the previous method is that now unsuccessful proposals are the ones that need to be rollbacked while successful ones can be discarded, while it was the other way in the first approach. Validity can be ensured at commit time but can not be ensured at rollback time, as a deallocated resource can have been fully allocated (commit and proposal discarded as it was considered valid without rollback) and so the rollback method would fail re-allocating that resource and thus unable to get to a valid state. A logical clock could also be used here but with the same implications that in the first case.
3. _Hybrid (allocate immediately and rollback later, prepare deallocate and commit later)_: this approach combines the previous two. It allocates the new resources at proposal preparation but doesn't release already in use resources until commit phase, when it either releases the deallocated resources if successful or the pre-allocated resources if not. As you can guess, both successful and unsuccessful proposals would need to be commited/rollbacked and none could be discarded, but the validity of the proposal can be ensured at commit time.

Summing up: (info about which operations are safe and which proposals can be dropped)
1. _Prepare a proposal and commit later_: Unsafe allocs and safe deallocs; commit successful and drop unsuccessful.
2. _Commit immediately and roll back later_: Safe allocs and unsafe deallocs; drop successful and rollback unsuccessful.
3. _Hybrid (allocate immediately and rollback later, prepare deallocate and commit later)_: Safe allocs and deallocs; commit successful and rollback unsuccessful.

The only method that is concurrency-proof is the third one. Its drawbacks are being unable to drop any proposal and reserving resources that would be locked while the proposal is active.

You claim that deallocs are not a problem since we can just roll forward so the second approach can be used. What do you mean by roll forward? The resource that we deallocated at proposal preparation may have been allocated by another proposal and thus we would be unable to get it back. Do you mean that we should allocate another resource and change it for the one we can not reallocate?